returnto path

pnoll1 opened the pull request:
adds returnto path for registration and sign in.  Appears to work correctly on test server.  The tests for:

- register an account captcha disabled
- we can create a new team
- we can create and edit a review
are failing.

eloquence commented on routes/helpers/render.js:
```diff
@@ -61,6 +61,8 @@ let render = {
     vars.urlPath = url.parse(req.originalUrl).pathname;
+    
+    vars.returnTo = req.get('Referer')
 
```
This won't work. Say I give you a link from mycoolwebsite.example to create an account on lib.reviews. If we just directly check the referer, you'll now be sent back to mycoolwebsite.example. Generally, the returnto should under no circumstances send the user back to a page that's not on lib.reviews itself. This is why sites like Wikipedia pass a relative returnto on as a request parameter on each page that links to the login or register button.

eloquence commented on routes/actions.js:
```diff
@@ -186,7 +186,7 @@ router.post('/signin', function(req, res, next) {
         return res.redirect('/signin');
       } else {
-        return res.redirect('/'); // Success
+        return res.redirect(req.body.returnto); // Success
```
The tests are failing because the referer on the signin and register page is .. the signin and register page. Try logging in by going directly to localhost/signin and you'll see that you can't. :) But we shouldn't be using the referer at all, more on that below.

pnoll1 commented:
implemented using querys in latest commit, please review methodology. Also, added logic to handle traffic going directly to signin and registration pages.

eloquence commented on routes/actions.js:
```diff
@@ -347,4 +348,10 @@ function sendRegistrationForm(req, res, formInfo) {
 
+function returnToPath(reqOrig, res) {
+    let reqUrl = reqOrig.originalUrl;
+    let query = url.parse(reqUrl, true).query.returnTo;
```
Is there a reason you are not just using req.query.returnTo here?

pnoll1 replied:
I thought req.query was incompatible with post requests due to some earlier issues. It's working now so I'll rewrite it to use that.

eloquence commented on routes/helpers/render.js:
```diff
@@ -61,7 +61,15 @@ let render = {
+        vars.returnTo = req.query.returnTo;
+    }
+    console.log('view', view);
+    console.log('returnTo',vars.returnTo);
```
These shouldn't be in the PR (we can add debug statements for stuff we really want to log, but this should not be necessary to log).

pnoll1 replied:
Noted.  These are personal debug statements I meant to delete.

eloquence commented on routes/helpers/render.js:
```diff
@@ -61,7 +61,15 @@ let render = {
     vars.qualifiedURL = config.qualifiedURL;
 
     vars.urlPath = url.parse(req.originalUrl).pathname;
```
I would suggest a separate variable for the path that's appended to the login/register links. Moreover, I would recommend that if there is already a returnTo in the query string, to use it as the value for that variable, rather than setting it here. I haven't tested that approach, but I think it would work.

The reason is that right now, if you go to /signin and _then_ go to /register (or click either button twice), it will _still_ pass along the "sign in" or "register" form as a return-to parameter, which is confusing. We should never, ever be redirected to a sign-in or register form after signing in or registering.

Consider the Wikipedia behavior:

1) Click https://en.wikipedia.org/wiki/Main_Page
2) Click the "Log in" link
3) Note how "returnto=Main+Page" is appended to the URL
4) Click the "Log in" link again
5) Note how "returnto=Main+Page" is _still_ appended to the URL
6) Click the "Create account" link
7) Note how "returnto=Main+Page" is _still_ appended to the URL

That is the behavior we should aim to reproduce.

pnoll1 replied:
This is what I aimed for but need to work out the cases for the if block.  I could check the path to not rewrite the returnTo on the signin and registration unless empty(from an external site).

The other way to handle the sign in/ registration page buttons would be to remove them.  They're redundant on these pages.

eloquence commented:
A lot nicer already! Still a couple outstanding issues, most importantly, we're still sometimes appending /signin or /register redirects. It also breaks a couple of tests, but we can fix that once those issues are addressed. See comments for details. :) Thanks again for working on this!

eloquence commented on routes/actions.js:
```diff
@@ -347,4 +348,10 @@ function sendRegistrationForm(req, res, formInfo) {
+function returnToPath(reqOrig, res) {
+    let reqUrl = reqOrig.originalUrl;
+    let query = url.parse(reqUrl, true).query.returnTo;
+    res.redirect(query);
```
This doesn't check against external vs. internal URL, so it could be used to redirect the user to another website. For security reasons, we have to prevent that, e.g., by ensuring that the returnTo value is a path.

eloquence commented on routes/actions.js:
```diff
@@ -347,4 +347,12 @@ function sendRegistrationForm(req, res, formInfo) {
+// checks for external redirect in returnTo, ignores if present and redirects
+function returnToPath(req, res) {
+    let returnTo = req.query.returnTo;
+    if (returnTo.includes('http')) {
```
This is not a good way to check for external URLs. First, there are many protocols -- http, https, ftp, mailto, etc. -- all of which we have to prevent redirects to. Second, checking anywhere within the string produces false positives such as "the-http-handbook". 

With your code, try http://localhost/signin?returnTo=//oops.com

That's what's known as a protocol-relative URL. The string http or https is not present in the URL; on lib.reviews it would redirect to https since lib.reviews itself is on https.

The approach I'd suggest is a regular expression that looks for a / at the beginning, followed by any character that is _not_ a slash. That seems like a reliable way to ensure that we're redirecting to a path on the same way.

The way to do this is something along these lines:

```javascript
const localPathRegex = new RegExp('^\/[^/]');
if (!localPathRegex.test(returnTo)) // Suppress redirect, could be a non-local resource
```

If you're unfamiliar with regular expressions,  you can use https://regex101.com/r/OxaMCS/1 to get an explanation of what this check does, and to perform your own sanity checks.

pnoll1 replied:
wouldn't changing the search string from http to // do the same thing?

Is there an advantage to using a regexp for this?

eloquence commented on routes/helpers/render.js:
```diff
@@ -62,6 +62,14 @@ let render = {
+    vars.returnTo = '/';
+    if (req.query.returnTo) {
+        vars.returnTo = req.query.returnTo;
+    } else if (req.path == 'signin' || req.path == 'register') { // do nothing
```
This condition won't be true, since the path will be /signin or /register (note leading slash), so you will instead move on to the `else` block. That's why right now http://localhost/signin (without returnTo) still redirects to itself. 

In addition to fixing that check, any reason not to move that initial `vars.returnTo = '/'` inside this `else if` block?

pnoll1 replied:
You're correct, I only tested it with querys present.

No, just how my brain was working that day.  I'll move it, the logic is readable if it's inside the block.

pnoll1 commented:
implemented external redirect check using .includes('//'), let me know if regexp is more appropriate for some reason.

Still failing 3-register an account(captcha disabled) with expected 302, got 200
going directly to register page and creating new user appears to work as intended during my testing

eloquence commented on routes/actions.js:
```diff
@@ -347,4 +347,12 @@ function sendRegistrationForm(req, res, formInfo) {
+// checks for external redirect in returnTo, ignores if present and redirects
+function returnToPath(req, res) {
+    let returnTo = req.query.returnTo;
+    if (returnTo.includes('//')) {
```
That test again won't be as reliable as the regex. Try, for example:

http://localhost/signin?returnTo=ftp:rtfm.mit.edu
http://localhost/signin?returnTo=mailto:some@address.examle (with mailto handler configured)

pnoll1 replied:
OK, I misunderstood your example.  We're trying to only allow paths that start with / since these will always be relative.  //, protocol:// and protocol: are all valid which means testing for them would be a fools errand and break often.

I'll implement the regexp.

pnoll1 commented:
all tests passing

eloquence commented:
Looks good! Merging, will make some minor cosmetic changes in a follow-up commit. Thank you again for your work on this. :)
